### ISSUE TITLE
Tracking - or detection of 'new' objects

### ISSUE THREAD

**Issue by simonbirt** [blakeblackshear/frigate#141]
I've been running Frigate 0.5.1-rc4 for a little while, with home assistant configured to alert when e.g. the binary sensor for car goes from 'off' to 'on'. This works pretty well for notifying us when a vehicle turns up at the gate. I'm now trying to work out how to suppress notifications for an object that remains in place for a while. E.g. the below scene - the car is parked. Every time the wind blows, the tree moves and I get a 'car detected' notification.
![image](https://user-images.githubusercontent.com/891658/83423551-0c931d80-a423-11ea-9c2d-9d9a6df21e3a.png)
This is presumably because the binary sensor is set to 'off' after a certain period of no motion. Is there any way to configure things so that a notification is only raised when the car first appears?

**Comment by blakeblackshear** [comment]
There isn't a good way to handle this at the moment. You could use a mask to look at only the areas you want notifications for.

**Comment by scstraus** [comment]
As this is pretty specific to your setup, it's probably not something that needs to be fixed in Frigate, but rather in your config. The way I'd solve this would be to make a template sensor that takes the state of the car detection sensor and when it turns off due to the tree blowing, it would keep it turned on until the car was re-detected. Just take the maximum amount of time that the car goes undetected by blowing trees for and make that the timeout value for when your template sensor finally decides that there's actually no car there any more.

**Comment by simonbirt** [comment]
Just to clarify - in the scene above the car was tracked as it arrived, but very shortly ( a few seconds) after parking it stopped being tracked. After that point, it's sporadically picked up as other motion is detected in the area. As far as I can see, the tracking should hold onto a detected object even if stationary. Is that not the case? I have the threshold for 'car' set at 0.8, so the detection above is comfortably over that.

**Comment by Buzztiger** [comment]
@blakeblackshear correct me if I'm wrong but I don't think frigate currently actually tracks objects. As far as I understand it's processed frame by frame first checking for motion between frames. If motion is detected only then the object recognition is started.

**Comment by scstraus** [comment]
> Just to clarify - in the scene above the car was tracked as it arrived, but very shortly ( a few seconds) after parking it stopped being tracked. After that point, it's sporadically picked up as other motion is detected in the area. As far as I can see, the tracking should hold onto a detected object even if stationary. Is that not the case? I have the threshold for 'car' set at 0.8, so the detection above is comfortably over that.

Okay that is different than how I understood it. In that case maybe it should be a frigate feature where we tell it to keep tracking for a certain amount of time. The side effect of this would likely be that it would find lots of these objects over a long period of time and eventually overload the coral with watching items that used to be there.. but maybe only for cars it could work if set to something reasonable like a day. I have to defer to Blake on that because I don’t understand how this works well enough.

**Comment by simonbirt** [comment]
Initially I thought tracking was being lost as objects temporarily dropped beneath the configured threshold - e.g. with a threshold of 0.8 tracking starts because a car is detected at 85%, but then stops if the confidence drops to 78%. 

I've patched this locally to allow the threshold to be proportionately reduced for frames that are coincident with tracked objects - i.e. if we detect a car at 85% and start tracking, then if in the next frame we detect a car in the same box we can allow a lower threshold (we know there was a car there in the last frame - it's very likely that it's still there). 

This does seem to provide more consistent tracking *for my situation*, but I'm still getting the case where stationary objects stop being tracked after a while, and I can't figure out why.

**Comment by simonbirt** [comment]
Tried this again with some additional logging. Drove a car into position and parked. The stationary vehicle was tracked as a car for some 780 frames. Then, while evaluating the same region, the detection came back as 'truck' for the next 10 frames with confidence varying between 50 and 73%. After 10 frames where no car was detected, tracking stopped (as expected). What I don't understand is why the detection would suddenly flip from 80% car to 50% truck for essentially the same image. I may try grouping car/truck as 'vehicle' for the purposes of tracking.

**Comment by blakeblackshear** [comment]
I see the same behavior. There was probably a slight change in lighting or background that caused the model to see a truck instead of a car. Grouping the car/truck labels under vehicle is exactly how I was planning to solve this problem. I am also planning to stop checking the same object when there is no motion after some point. No sense in wasting detection cycles looking again and again at something if nothing has moved.

**Comment by niemyjski** [comment]
Agreed

**Comment by scstraus** [comment]
> I am also planning to stop checking the same object when there is no motion after some point. No sense in wasting detection cycles looking again and again at something if nothing has moved.

Would be good if this was configurable ie how many frames to keep checking before you give up.

**Comment by blakeblackshear** [comment]
I think you misinterpreted. It won't give up. It will assume the object is still there unless there is motion in the bounding box.

**Comment by scstraus** [comment]
I guess I'm misunderstanding the whole thread ;-). Well, that's a harder problem. Can detection regions overlap? Maybe there could be a configurable number of frames that it will look in the same place for the same object again if new motion was detected in the same area, so it looks for the new thing and the old thing again for some time to figure out if the old one is still there while also looking for the new one..?

**Comment by niemyjski** [comment]
What if it’s because you decrement the % too much that it hits a lower threashold after n>50 times. Maybe set a min amount.

**Comment by simonbirt** [comment]
I wasn't planning to raise a PR as I expect you'll be doing this rather differently, especially when you implement the change to stop checking stationary objects. However if it's of interest [these changes](https://github.com/blakeblackshear/frigate/compare/master...simonbirt:track_alias) are what seem to be working for me at the moment. This implements:

- Aliasing for model labels, allowing labels to be grouped
- An optional reduced threshold once an object is acquired for tracking. The reduction is applied in proportion to the percentage overlap of the current detection box to the previous.
- A size limit to the history list to prevent it growing unchecked.

**Comment by simonbirt** [comment]
And now the Early Warning of Escaped Sheep (EWES) is up and running :-)

<img width="509" alt="image" src="https://user-images.githubusercontent.com/891658/84024467-d9e8a800-a981-11ea-83ae-4f1d989b316d.png">

**Comment by Buzztiger** [comment]
Lol, is the system automatically deploying active countermeasures and effectors, for example the D.O.G. device?

**Comment by valantur** [comment]
> Lol, is the system automatically deploying active countermeasures and effectors, for example the D.O.G. device?

You didn't get enough praise for that joke. I'm here to correct that.

**Comment by idittansikte** [comment]
I've replaced the current tracking algorithm with SORT which have some useful config for handling this problem.

I made that change since I've modifed frigate to be able to track where people are in my flat with pretty good results. (I have a central corridor where my camera is at.)

I am planning on uploading the code (either as a pull request or on the side of this repo) as soon when I get some time cleaning it up a bit. If someone is interested I can hurry.
